packages/mobile/android/app/src/main/java/com/quietmobile/Backend/TorHandler.java

```diff
@@ -0,0 +1,310 @@
+                    ));
+                }
+
+                String[] verifyLines = lines.toArray(new String[0]);
+                if (!mainConfigurationSetCommandLine(verifyLines)) {
+                    throw new IllegalArgumentException("Setting command line failed: " + Arrays.toString(verifyLines));
+                }
+
+                int result = runMain(); // run verify
+                if (result != 0) {
+                    throw new IllegalArgumentException("Bad command flags: " + Arrays.toString(verifyLines));
+                }
+
+                if (ENABLE_CONTROL_SOCKET) {
+                    controlPortThreadStarted = new CountDownLatch(1);
+                    controlPortThread.start();
+                    controlPortThreadStarted.await();
+                }
+
+                String[] runLines = new String[lines.size() - 1];
+                runLines[0] = "tor";
+                for (int i = 2; i < lines.size(); i++) {
+                    runLines[i - 1] = lines.get(i);
+                }
+                if (!mainConfigurationSetCommandLine(runLines)) {
+                    throw new IllegalArgumentException("Setting command line failed: " + Arrays.toString(runLines));
+                }
+                if (ENABLE_CONTROL_SOCKET && !mainConfigurationSetupControlSocket()) {
+                    throw new IllegalStateException("Setting up ControlPort failed!");
+                }
+                if (runMain() != 0) {
+                    throw new IllegalStateException("Tor could not start!");
+                }
+
+            } catch (IllegalStateException | IllegalArgumentException | InterruptedException e) {
+                e.printStackTrace();
+            } finally {
+                mainConfigurationFree();
+                runLock.unlock();
+            }
+        }
+    };
+
+    /**
+     * Start Tor in a {@link Thread} with the minimum required config.  The
+     * rest of the config should happen via the Control Port.  First Tor
+     * runs with {@code --verify-config} to check the command line flags and
+     * any {@code torrc} config. Then finally Tor is
+     * started in its own {@code Thread}.
+     * <p>
+     * Tor daemon does not output early debug messages to logcat, only after it
+     * tries to connect to the ports.  So it is important that Tor does not run
+     * into port conflicts when first starting.
+     *
+     * @see <a href="https://trac.torproject.org/projects/tor/ticket/32036">#32036  output debug logs to logcat as early as possible on Android</a>
+     * @see <a href="https://github.com/torproject/tor/blob/40be20d542a83359ea480bbaa28380b4137c88b2/src/app/config/config.c#L4730">options that must be on the command line</a>
+     */
+    public void startTorThread() {
+        if (runLock.isLocked()) {
+            Log.i(TAG, "Waiting for lock");
+        }
+        runLock.lock();
+        Log.i(TAG, "Acquired lock");
+        torThread.start();
+    }
+
+}
```

Comment:
Is this code borrowed from somewhere?

Reply:
I based this class on guardian project's [TorService.js](https://github.com/guardianproject/tor-android/blob/master/tor-android-binary/src/main/java/org/torproject/jni/TorService.java) yet it's modified by myself: I  transformed it from Service into Handler, made it configurable to use control socket and enabled control port as default control mechanism (with it's configuration passed down from the parent class) secured with authentication cookie

Reply:
It is, right @siepra? We need to discuss how to organize this so that the license is clear.